_data/build-research.py

```diff
@@ -1,85 +1,209 @@
-    new_paper["date"] = datetime.strptime(new_paper["date"], "%Y-%m-%d").strftime("%Y-%m-%d")
+    log("------------------------------------------------------------")
+    log("")
+    log(f"Paper {input_index + 1} of {len(input_papers)}")
+    log(f"Line number {input_paper.get('__line_number__')}")
+
+    # catch errors
+    try:
+
+        # check that paper entry is dictionary/object
+        if type(input_paper) != dict:
+            raise Exception("Entry not in expected format. Skipping.")
+
+        # get paper entry identifier
+        input_id = input_paper.get("id")
+        if not input_id:
+            raise Exception("Entry has no id field. Skipping.")
+        log(input_id)
+        log("")
+
+        # find duplicate entries
+        for duplicate_index, duplicate_paper in enumerate(input_papers):
+            if duplicate_paper.get("id") == input_id:
+                # don't count first instance of id as duplicate
+                if input_index > duplicate_index:
+                    raise Exception("Entry is a duplicate. Skipping.")
+
+        # find same paper in existing output
+        existing_paper = None
+        for output_paper in output_papers:
+            if output_paper.get("id") == input_id:
+                existing_paper = output_paper
+                break
+
+        # if already in output, use existing paper to save time
+        if existing_paper:
+            log("Already in output. Using existing citation.", "blue")
+            log("")
+            new_papers.append(existing_paper)
+
+        # if not already in output, run Manubot to get citation info
+        else:
+            log("Running Manubot to generate citation")
+            log("")
+
+            # run Manubot and get results as json
+            try:
+                commands = ["manubot", "cite", input_id, '--log-level=ERROR']
+                output = subprocess.Popen(commands, stdout=subprocess.PIPE)
+                citation = json.loads(output.communicate()[0])[0]
+            except Exception:
+                log("")
+                raise Exception("Manubot could not generate citation")
+
+            # new paper info, with only needed info from citation
+            new_paper = {}
+
+            # title
+            new_paper["title"] = citation.get("title", "")
+
+            # authors
+            new_paper["authors"] = []
+            for author in citation.get("author", []):
+                given = author.get("given", "")
+                family = author.get("family", "")
+                new_paper["authors"].append(given + " " + family)
+
+            # publisher
+            container = citation.get("container-title", "")
+            collection = citation.get("collection-title", "")
+            publisher = citation.get("publisher", "")
+            new_paper["publisher"] = container or publisher or collection
+
+            # date
+            def part(index):
```

Comment:
You may want to define `part(..)` as a global function instead of redefining it in each iteration of the `for` loop.